dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java

```diff
@@ -562,54 +551,63 @@ public Workflow upsertServiceVersion(@ApiParam(hidden = true) @Auth User user,
-            KeyFactory keyFactory = KeyFactory.getInstance("RSA");
-            PKCS8EncodedKeySpec pkcs8EncodedKeySpec = new PKCS8EncodedKeySpec(Base64.getDecoder().decode(pemFileContent));
-            rsaPrivateKey = (RSAPrivateKey)keyFactory.generatePrivate(pkcs8EncodedKeySpec);
-        } catch (IOException | NoSuchAlgorithmException | InvalidKeySpecException ex) {
-            LOG.error(ex.getMessage(), ex);
-        }
-
-        if (rsaPrivateKey != null) {
-            final int tenMinutes = 600000;
-            try {
-                Algorithm algorithm = Algorithm.RSA256(null, rsaPrivateKey);
-                String jsonWebToken = JWT.create()
-                        .withIssuer(gitHubAppId)
-                        .withIssuedAt(new Date())
-                        .withExpiresAt(new Date(Calendar.getInstance().getTimeInMillis() + tenMinutes))
-                        .sign(algorithm);
-                CacheConfigManager.setJsonWebToken(jsonWebToken);
-            } catch (JWTCreationException ex) {
-                LOG.error(ex.getMessage(), ex);
-            }
-        }
+    private void syncServices(User user, Optional<String> organization, Token gitHubToken) {
+        GitHubSourceCodeRepo gitHubSourceCodeRepo = (GitHubSourceCodeRepo)SourceCodeRepoFactory.createSourceCodeRepo(gitHubToken, client);
+
+        // Get all GitHub repositories for the user
+        final Map<String, String> workflowGitUrl2Name = gitHubSourceCodeRepo.getWorkflowGitUrl2RepositoryId();
+
+        // Filter by organization if necessary
+        final Collection<String> repositories = GitHubHelper.filterReposByOrg(workflowGitUrl2Name.values(), organization);
+
+        // Add user to any services they should have access to that already exist on Dockstore
+        final List<Workflow> existingWorkflows = findDockstoreWorkflowsForGitHubRepos(repositories);
+        existingWorkflows.stream()
+                .filter(workflow -> !workflow.getUsers().contains(user))
+                .forEach(workflow -> workflow.getUsers().add(user));
+        final Set<String> existingWorkflowPaths = existingWorkflows.stream()
+                .map(workflow -> workflow.getWorkflowPath()).collect(Collectors.toSet());
+
+        GitHubHelper.checkJWT(gitHubAppId, gitHubPrivateKeyFile);
+
+        GitHubHelper.reposToCreateServicesFor(repositories, organization, existingWorkflowPaths).stream()
+                .forEach(repositoryName -> {
+                    final Service service = gitHubSourceCodeRepo.initializeService(repositoryName);
+                    service.addUser(user);
+                    final long serviceId = workflowDAO.create(service);
+                    final Workflow createdService = workflowDAO.findById(serviceId);
+                    final Workflow updatedService = gitHubSourceCodeRepo.getWorkflow(repositoryName, Optional.of(createdService));
+                    updateDBWorkflowWithSourceControlWorkflow(createdService, updatedService);
```

Comment:
This going a bit too far actually, I think this pulls down all versions and their files which actually is a refresh. 
Probably a miscommunication, but I was thinking the following procedure
1) A user logs in
2) A user wants to inform us about their repos that they belong to that have had the app installed by someone else. 
3) The user hits sync which is speedy and simply adds them to the services that they should have access to (like above with line 586). 
4) They examine the resulting services and only refreshes the individual services/workflows which may have problems (say buggy webhook delivery).

Reply:
This is will also handle this case; in fact it's the main case I thought I was supposed to be solving:

1. A user creates a Dockstore service in GitHub and a GitHub release on a repo
2. The user then installs the Dockstore GitHub app on the repo.
3. This code ensures that a Dockstore version is created for the release that existed before installing the app.

I think we're on the same page, but just making sure.

I think we are on the same page, but just making sure.

Reply:
Yup, that makes sense, I was just working out the implications of the change.